python/GangaDirac/Lib/Utilities/DiracUtilities.py

```diff
@@ -31,36 +29,79 @@
 def getDiracEnv(force=False):
     """
     Returns the dirac environment stored in a global dictionary by Ganga.
-    This is expected to be stored as some form of 'source'-able file on disk which can be used to get the printenv after sourcing
-    Once loaded and stored his is used for executing all DIRAC code in future
+    Once loaded and stored this is used for executing all DIRAC code in future
+
     Args:
         force (bool): This triggers a compulsory reload of the env from disk
     """
     global DIRAC_ENV
     with Dirac_Env_Lock:
-        if DIRAC_ENV == {} or force:
-            config_file = getConfig('DIRAC')['DiracEnvFile']
-            if not os.path.exists(config_file):
-                absolute_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), '../../..', config_file)
+        if not DIRAC_ENV or force:
+            cache_file = getConfig('DIRAC')['DiracEnvJSON']
+            if cache_file:
+                DIRAC_ENV = read_env_cache(cache_file)
             else:
-                absolute_path = config_file
-            if getConfig('DIRAC')['DiracEnvFile'] != "" and os.path.exists(absolute_path):
+                source_command = getConfig('DIRAC')['DiracEnvSource']
+                if source_command:
+                    DIRAC_ENV = get_env(source_command)
+                else:
+                    logger.error("'DiracEnvSource' config variable empty")
+    return DIRAC_ENV
 
-                env_dict = {}
-                logger.debug("Executing command: %s" % 'source {0}'.format(absolute_path))
-                execute('source {0}'.format(absolute_path), shell=True, env=env_dict, update_env=True)
 
-                if env_dict is not None:
-                    DIRAC_ENV = env_dict
-                else:
-                    logger.error("Error determining DIRAC environment")
-                    raise GangaException("Error determining DIRAC environment")
+def get_env(env_source):
+    """
+    Given a source command, return the DIRAC environment that the
+    command created.
 
-            else:
-                logger.error("'DiracEnvFile' config variable empty or file not present")
-                logger.error("Tried looking in : '%s' Please check your config" % absolute_path) 
-    logger.debug("Dirac Env: %s" % DIRAC_ENV)
-    return DIRAC_ENV
+    Args:
+        env_source: a command which can be sourced, providing the desired environment
+
+    Returns:
+        dict: the environment
+
+    """
+    logger.debug('Running DIRAC source command %s', env_source)
+    env = dict(copy(os.environ))
```

Comment:
I see your point now, I think I was confusing myself. So, the `copy` was needed so that we don't change the global variable. The `dict()` is needed since `os.environ` is not a Python `dict` but rather of type `os._Environ`. You are right though that both are not needed, simply converting it to a `dict()` should be sufficient. I will update this.